### PR TITLE
Added class file for delaws

### PR DESCRIPTION
**Pull request by lflage**
The file contain methods taken from the [de_laws_to_json](https://github.com/bundestag/de_laws_to_json/tree/main) repository to ensure the quality of the data.

I had to overwrite the `download()` method because wget was not working properly and added a custom request function that I use in general to deal with timeout problems.

The raw dataset is saved as `xml` and the `process_text()` parses the data removing duplicates and ordering the paragraphs.

### CODE REVIEW

File: src/lm_datasets/datasets/de/de_laws.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    Class to download, process and yield a dataset of german law.
+    Parts of code adapted from:
+    @bundestag
+    https://github.com/bundestag/de_laws_to_json#Apache-2.0-1-ov-file
+    """
+
+    DATASET_ID = "delaws"           # TODO: Check Naming conventions
+    TITLE = "DELaws"
+    DESCRIPTION = (
+        "Crawled documents from the DeLaws repository"
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:50]: Are the texts extracted from gesetze-im-internet.de?

File: src/lm_datasets/datasets/de/de_laws.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+from tqdm import tqdm
+from typing import Dict, Union, Optional
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+from lm_datasets.datasets.base import BaseDataset, Availability, MILLION
+
+logger = logging.getLogger(__name__)
+
+
+def custom_request(url):
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:23]: Instead of retrying, I recommend adding a delay between the request.

File: src/lm_datasets/datasets/de/de_laws.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        """
+        if element.string:
+            return element.string
+        else:
+            children_dict = {}
+            for child in element.contents:
+                if child.name:
+                    if child.name in children_dict:
+                        if isinstance(children_dict[child.name], list):
+                            children_dict[child.name].append(DELawsDataset.convert_xml_to_dict(child))
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:101]: better use "self" to call class methods

File: src/lm_datasets/datasets/de/de_laws.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+    def num_tokens_from_string(string: str) -> int:
+        """
+        Function to count the number of tokens in a string
+        https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
+        """
+        encoding = tiktoken.get_encoding('cl100k_base')
+        num_tokens = len(encoding.encode(string))
+        return num_tokens
+
+    def convert_xml_to_dict(element, expected_type: Optional[type] = None) -> Union[str, Dict]:
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:89]: Make this a class method with self or static

File: src/lm_datasets/datasets/de/de_laws.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+            item_dict = {'title': title, 'link': link, 'output_dir': self.output_dir}
+            item_array.append(item_dict)
+
+        # Create directory if it doesn't exist
+        if not os.path.exists(self.output_dir):
+            os.makedirs(self.output_dir)
+        
+        # Set the number of items to process
+        num_items_to_process = len(item_array)  # change this to control how many items to process
+        print(f"Processing {num_items_to_process} items out of {len(item_array)} total items")
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:137]: Please use "logger.info" (or debug) instead of print

File: src/lm_datasets/datasets/de/de_laws.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+        # Initialize a Pool with the number of available processors
+        pool = multiprocessing.Pool(processes=multiprocessing.cpu_count())
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:140]: Don't use all CPU cores, the BaseDataset has the worker attribute for this.

File: src/lm_datasets/datasets/de/de_laws.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+        print(f"Using {multiprocessing.cpu_count()} cores/processes in parallel")
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:141]: -> logger

File: src/lm_datasets/datasets/de/de_laws.py [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+
+        # Use Pool's map function to process the items in parallel
+        with tqdm(total=num_items_to_process, desc="Processing files", dynamic_ncols=True) as pbar:
+            for _ in pool.imap_unordered(DELawsDataset.process_law, item_array[:num_items_to_process]):
+                pbar.update()
+    
+    def process_text(self, filename):
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:148]: Maybe "extract_text_from_xml_file"? "process_text" is pretty vague.

File: src/lm_datasets/datasets/de/de_laws.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+                text = ''
+                for norm in value:
+                    text += norm['meta']['title']
+                    text += '\n'
+                    for content in norm['paragraphs']:
+                        text += content['content']
+                        text += '\n'
+            return text
+
+    def get_text(self):
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:324]: It should be "get_texts"

File: src/lm_datasets/datasets/en/edgar_corpus.py
Comment (malteos) [inline src/lm_datasets/datasets/en/edgar_corpus.py:1]: --> move to different PR

File: src/lm_datasets/datasets/multilingual/wura.py
Comment (malteos) [inline src/lm_datasets/datasets/multilingual/wura.py:1]: --> move to different PR

**Review by malteos — CHANGES_REQUESTED** [review]
Thanks for the PR.

You also need to add the dataset class to the registry (https://github.com/malteos/lm-datasets/blob/main/src/lm_datasets/datasets/dataset_registry.py).

For everything else see the inline comments.

File: src/lm_datasets/datasets/de/de_laws.py [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+
+    DATASET_ID = "delaws"           # TODO: Check Naming conventions
+    TITLE = "DELaws"
+    DESCRIPTION = (
+        "Crawled documents from the DeLaws repository"
+    )
+    LANGUAGES = ['de']
+    AVAILABILITY = Availability.DIRECT_DOWNLOAD
+    WEB_CRAWLED = True
+    # LICENSE =                     # TODO:
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:55]: Laws are considered "public domain" in Germany.

File: src/lm_datasets/datasets/de/de_laws.py [hunk 11]
```diff
@@ -0,0 +1,334 @@
+import os
+import zipfile 
+import re
+import multiprocessing
+import requests
+import time
+import logging
+import copy
+import pprint
+import tiktoken
```
Comment (malteos) [inline src/lm_datasets/datasets/de/de_laws.py:10]: Is this dependency necessary?

File: src/lm_datasets/datasets/de/de_laws.py (hunk 2 above)
Comment (lflage) [reply src/lm_datasets/datasets/de/de_laws.py:23]: You mean, simply putting a timer between requests if the request failed?
Or a delay between every request?

Comment (malteos) [reply src/lm_datasets/datasets/de/de_laws.py:22]: Rather add a delay between every request. The other approach will annoy the site you are scraping.

**Review by malteos — APPROVED** [review]
Thanks!
